Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -166,19 +199,32 @@ public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
         guard let firstLine = ByteString(splitted[0]).validDescription,
               let match = ToolsVersionLoader.regex.firstMatch(
                   in: firstLine, options: [], range: NSRange(location: 0, length: firstLine.count)),
-              match.numberOfRanges >= 2 else {
-            return (nil, bytes.contents)
+              // The 3 ranges are:
+              //   1. The entire matched string.
+              //   2. Capture group 1: the comment spacing.
+              //   3. Capture group 2: The version specifier.
+              // Since the version specifier is in the last range, if the number of ranges is less than 3, then no version specifier is captured by the regex.
+              // FIXME: Should this be `== 3` instead?
```

Comment:
It makes sense to use `>=` but then maybe have a `assert(numRanges == 3)` in the if-condition, to make sure it's updated if somebody changes the regex in the future.  But I think it would also be fine in this case to just leave it at `>= 3`.

Reply:
Sorry for my late response to this. I'm planning on removing the regex altogether, in favour of working directly with `String` and `Substring`. In my opinion, regex is harder to understand and maintain. Also, working with `NSRegularExpression` means `String` gets bridged to and from `NSString` many times, which is inefficient compared to working with just `String` and `Substring`.

Reply:
I started to phase out regular expressions in b370960f12434ec523a84bb6856ea773842edb2a, and restored `regex` to what it's like on the main branch (except for some extra comments). My intention is to remove it completely once `ToolsVersionLoader.split(_: ByteString)` is removed. Do you think it should still use `>=` and `assert`?

Reply:
I agree with you about transitioning from regex here.  Given that, I don't think this matters.  I would just remove the FIXME.

Reply:
Sorry I forgot that the entire `regex` along with the old `split(_:)` were already removed in 06da1ded449207717396c76f73d3322c2995778e, so the FIXME was already removed as well.